src/cockpit/channel.py

```diff
@@ -167,8 +180,48 @@ def ready(self, **kwargs):
     def done(self):
         self.send_control(command='done')
 
+    # tasks and close management
+    def is_closing(self) -> bool:
+        return self._close_args is not None
+
+    def _close_now(self):
+        self.send_control('close', **self._close_args)
+
+    def _task_done(self, task):
+        # Strictly speaking, we should read the result and check for exceptions but:
+        #   - exceptions bubbling out of the task are programming errors
+        #   - the only thing we'd do with it anyway, is to show it
+        #   - Python already does that with its "Task exception was never retrieved" messages
```

Comment:
[This failure](https://cockpit-logs.us-east-1.linodeobjects.com/pull-18356-20230216-094601-9474f71d-fedora-37-pybridge/log.html#72) is an interesting case study for this comment/claim:
```
asyncio-ERROR: Task exception was never retrieved
systemd_ctypes.bus.BusError: org.freedesktop.DBus.Error.AccessDenied: Failed to get file context on /usr/lib/systemd/system/NetworkManager.service.
```
This actually kind of *is* an error which should be caught and bubbled up through the protocol as `access-denied`.

This failure is *not* new -- since we landed #18321 we've been getting this on some test runs. It's not a regression really, #18321 just made these errors actually visible. In fact, in a run there we got the exact same test failure/message.

So TL/DR: I think we can ignore/retry the test run for this particular case, but adding this general claim as a comment here seems wrong?

Reply:
In other words, can we actually collect the task error, close the channel with an appropriate problem code for `PermissionDeniedError`, D-Bus errors and similar run-time/user errors (what are these, `OSError`?), and log the crash otherwise?

Reply:
That said, as (1) this isn't new, and (2) we have a constant reminder that we need to clean this up through failed tests, I don't veto this. Everything else looks good to me, and I'm happy to see this landing to untangle this web.

Reply:
However, let's please retry and only land if green. It's very plausible that the effect of this PR is to turn this failure from "flaky" to "100% reliable", as we now wait for tasks to close. Otherwise, if it repeats, we at least have a reliable way for reproducing this scenario, which is hugely helpful as well.

Reply:
I consider the claim in this comment to be 100% correct to this case.  Specifically: exceptions bubbling out of the task are programming errors.  This particular failure is a programming error and we need to fix it.  The fact that Python has shown this to us helps us fix this error.

This is exactly what I had in mind when I wrote that comment.

Reply:
Bug for this → https://github.com/cockpit-project/cockpit/issues/18358